index.ts

```diff
@@ -119,7 +116,7 @@ async function handleClick(
 
 		throw error;
 	} finally {
-		void updateItem();
+		void updateItem(tab?.id ? await getTabUrl(tab?.id) : '');
```

Comment:
Rather, I suppose:

- restore the code on line 76
- also add the same exact code before `throw error;`

right?

Reply:
> If this only needs to be called on error, then it should be inside the catch statement right?

I did think about that, but I wanted to be able to add handling for other scenarios like this one:

https://github.com/aspiers/webext-domain-permission-toggle/commit/412c07089ca95efdf4573adb892b72dc9c72ad91

and so it felt like a cleaner separation of concerns to only update the context menu from `updateItem()` and then have that function be smart enough to handle all scenarios, rather than having multiple parts of the code updating the menu which could potentially interact in unpredictable and undesirable ways (e.g. race conditions from having multiple event handlers collide).

That said, this is just a preference rather than a strong opinion.

> Also I presume that if there's an error the status is "unchecked", which doesn't need a URL check.

I wasn't sure about that.  Couldn't the extension already be enabled, in which case if there's an error it should stay checked?

> On extension with limited permissions, you can't get the tab URL (e.g. I think this fails on chrome://extensions for example)

There are so many pitfalls in this area it's crazy 🤣

> There's a demo extension in this repo you can use for testing. Use `npm run demo:watch` to build it.

Oh thanks, I'll check it out!

> Rather, I suppose:
> 
> * restore the code on line 76
> * also add the same exact code before `throw error;`
> 
> right?

You mean in addition to removing the `finally` clause?  Otherwise there could be duplicate menu updates: in both the `catch` and `finally` clauses.

Reply:
> I wasn't sure about that. Couldn't the extension already be enabled, in which case if there's an error it should stay checked?

There shouldn't be an error when removing a permission. However… you can't remove a wide permission like `*://*/*` this way. If that optional permission is granted, the item should become "✔️ Enabled on all sites". Clicking it would remove that wider permission. 

Maybe open a new issue. It's an edge case.

Reply:
OK, new version coming shortly.

Reply:
TL;DR: after investigation, I recommend keeping the approach in this PR, but the gory details follow.

So I tried this approach:

https://github.com/aspiers/webext-domain-permission-toggle/commit/bb77a79b6695df48159c001af936b3a3ba5219b1

(ignore the outdated commit message which I didn't update) and it results in several regressions where the check mark doesn't get set correctly in multiple cases.

For example, switch site access to on click, then enable for a site in the manifest via content menu toggle.  Because we're relying on Chrome to update the menu item, it doesn't spot that the site is in the manifest, therefore it (correctly) enables the checkmark but fails to disable (grey out) the menu item.  This means that it's now possible to attempt to remove the same permission from the still enabled menu item.  In that case, this statement:

> There shouldn't be an error when removing a permission.

turns out to be not true, because it results in:

    Uncaught (in promise) Error: You cannot remove required permissions.

Even if we add some logic to deal with that case, it feels to me like a very brittle approach to assume that we can ensure `chrome.permissions.remove()` will only ever be called in a way to ensure its success, because there are so many different combinations of scenarios which we don't fully understand yet - even in Chrome, let alone other browsers.

As well as being brittle, it would basically require either:

- copying logic into `togglePermission()` from `updateItem()` which tests URLs against the manifest, which would be a violation of DRY and separation of concerns, or
- calling `updateItem()` from `togglePermissions()`, in which case it's not really any different to just keeping `updateItem()` in the `finally` block. 

IMHO this affirms the point I was making earlier about the benefit of separation of concerns: if we have one function which handles interactions with `chrome.permissions` (i.e. `togglePermission()`, and then a separate function which takes care of ensuring that the menu item is always configured correctly (`updateItem()`), then we don't really need to worry about all the ways they can interact with each other.  All we need to do is ensure that `updateItem()` is called at the appropriate points.

Revisiting your original comments with the benefit of hindsight:

> If this only needs to be called on error,

We now know this isn't true, because of the need to disable the menu item in certain cases.

> Also I presume that if there's an error the status is "unchecked", which doesn't need a URL check.

We now know this isn't true, since if something prior to this goes wrong, it can allow a misguided attempt to disable a site in the manifest would fail, in which case the status should remain checked.

> On extension with limited permissions, you can't get the tab URL (e.g. I think this fails on chrome://extensions for example)

I don't see a problem here, because the code in this PR should still work if the tab URL can't be retrieved.  The only minor issue might be that `updateItem()` will cause the menu item to be enabled, which doesn't make sense for things like `chrome://extensions`.  But that could be easily fixed, e.g. by changing the default to `enabled: false`.

Reply:
> But that could be easily fixed, e.g. by changing the default to `enabled: false`.

Nothing is "easily" fixed here obviously haha


> > If this only needs to be called on error,
> 
> We now know this isn't true, because of the need to disable the menu item in certain cases.

I'm only talking about `updateItem` calls in `action.onClicked`, not `tabs.onActivated`


> and it results in several regressions where the check mark doesn't get set correctly in multiple cases.

As mentioned elsewhere, I really don't want to get into `allow access on click` and other Chrome UI stuff. Let's fix this first.

Reply:
I'll open a PR